presto-rcfile/src/main/java/com/facebook/presto/rcfile/RcFileReader.java

```diff
@@ -0,0 +1,584 @@
+
+    public long getReadTimeNanos()
+    {
+        return dataSource.getReadTimeNanos();
+    }
+
+    public Slice getSync()
+    {
+        Slice sync = Slices.allocate(SIZE_OF_LONG + SIZE_OF_LONG);
+        sync.setLong(0, syncFirst);
+        sync.setLong(SIZE_OF_LONG, syncSecond);
+        return sync;
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        blockPosition = 0;
+        blockRowCount = 0;
+        currentChunkSize = 0;
+        input.close();
+    }
+
+    public int nextBlock()
+            throws IOException
+    {
+        blockPosition += ColumnData.MAX_SIZE;
+        currentChunkSize = min(ColumnData.MAX_SIZE, blockRowCount - blockPosition);
+
+        // do we still have rows in the current block
+        if (currentChunkSize > 0) {
+            return currentChunkSize;
+        }
+
+        // do we have a next block
+        if (input.position() >= end || input.remaining() < SIZE_OF_INT) {
+            close();
+            return -1;
+        }
+
+        // read uncompressed size of block (which is useless information)
+        int uncompressedBlockSize = Integer.reverseBytes(input.readInt());
+
+        // read sequence sync if present
+        if (uncompressedBlockSize == -1) {
+            if (input.remaining() < SIZE_OF_LONG + SIZE_OF_LONG + SIZE_OF_INT) {
+                close();
+                return -1;
+            }
+            if (syncFirst != input.readLong() || syncSecond != input.readLong()) {
+                throw corrupt("Invalid sync in RCFile %s", dataSource);
+            }
+
+            // read the useless uncompressed length
+            uncompressedBlockSize = Integer.reverseBytes(input.readInt());
+        }
+        if (uncompressedBlockSize <= 0) {
+            throw corrupt("Invalid uncompressed block length %s", uncompressedBlockSize);
+        }
+
+        // read block header
+        int uncompressedHeaderSize = Integer.reverseBytes(input.readInt());
+        int compressedHeaderSize = Integer.reverseBytes(input.readInt());
+        if (codec == null && compressedHeaderSize != uncompressedHeaderSize) {
+            close();
+            throw corrupt("Invalid RCFile %s", dataSource);
+        }
+        if (compressedHeaderSize > compressedHeaderBuffer.length()) {
+            compressedHeaderBuffer = Slices.allocate(compressedHeaderSize);
+        }
+        input.readBytes(compressedHeaderBuffer, 0, compressedHeaderSize);
+
+        // decompress block header
+        Slice header;
+        if (codec != null) {
+            if (headerBuffer.length() < uncompressedHeaderSize) {
+                headerBuffer = Slices.allocate(uncompressedHeaderSize);
+            }
+            Slice buffer = headerBuffer.slice(0, uncompressedHeaderSize);
+
+            codec.decompress(compressedHeaderBuffer, buffer);
+
+            header = buffer;
+        }
+        else {
+            header = compressedHeaderBuffer.slice(0, uncompressedHeaderSize);
```

Comment:
Remove the `.slice(...)` part.

Maybe move the `compressedHeaderSize != uncompressedHeaderSize` sanity check down here to readability.